File: android/src/main/java/com/fitnessgeolocation/LocationDatabase.kt
package com.fitnessgeolocation

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log

class LocationDatabase(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, 6) {
  companion object {
    private const val DB_NAME = "fitness_geolocation.db"
    private const val TAG = "FitnessGeoDB"
  }

  override fun onCreate(db: SQLiteDatabase) {
    db.execSQL("""
      CREATE TABLE IF NOT EXISTS locations (
        id TEXT PRIMARY KEY,
        latitude REAL NOT NULL,
        longitude REAL NOT NULL,
        accuracy REAL,
        speed REAL,
        heading REAL,
        altitude REAL,
        timestamp INTEGER NOT NULL,
        battery_level REAL DEFAULT -1,
        motion_state TEXT DEFAULT 'unknown',
        signal_strength TEXT DEFAULT 'medium',
        session_id TEXT,
        delivered_to_js INTEGER DEFAULT 0,
        distance_from_prev REAL DEFAULT 0,
        cumulative_distance REAL DEFAULT 0,
        quality_json TEXT
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_locations_pending ON locations(delivered_to_js)")
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_locations_session ON locations(session_id)")
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_locations_timestamp ON locations(timestamp)")

    db.execSQL("""
      CREATE TABLE IF NOT EXISTS sessions (
        id TEXT PRIMARY KEY,
        name TEXT,
        activity_type TEXT DEFAULT 'running',
        start_time INTEGER NOT NULL,
        end_time INTEGER,
        total_distance REAL DEFAULT 0,
        total_duration INTEGER DEFAULT 0,
        total_active_duration INTEGER DEFAULT 0,
        max_speed REAL DEFAULT 0,
        elevation_gain REAL DEFAULT 0,
        average_accuracy REAL DEFAULT 0,
        point_count INTEGER DEFAULT 0,
        pause_count INTEGER DEFAULT 0,
        uploaded INTEGER DEFAULT 0,
        extras TEXT
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_sessions_uploaded ON sessions(uploaded)")

    // ─── v5: Engine event stores (offline-first) ─────────────────────────────
    db.execSQL("""
      CREATE TABLE IF NOT EXISTS diagnostics_timeline (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        session_id TEXT,
        level TEXT NOT NULL,
        event TEXT NOT NULL,
        data_json TEXT,
        timestamp INTEGER NOT NULL
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_diag_session_ts ON diagnostics_timeline(session_id, timestamp)")

    db.execSQL("""
      CREATE TABLE IF NOT EXISTS motion_events (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        session_id TEXT,
        activity TEXT NOT NULL,
        confidence REAL DEFAULT 0,
        data_json TEXT,
        timestamp INTEGER NOT NULL
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_motion_session_ts ON motion_events(session_id, timestamp)")

    db.execSQL("""
      CREATE TABLE IF NOT EXISTS geofence_events (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        session_id TEXT,
        identifier TEXT NOT NULL,
        action TEXT NOT NULL,
        data_json TEXT,
        timestamp INTEGER NOT NULL
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_geofence_session_ts ON geofence_events(session_id, timestamp)")

    db.execSQL("""
      CREATE TABLE IF NOT EXISTS sync_queue (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        kind TEXT NOT NULL,
        payload_json TEXT NOT NULL,
        attempts INTEGER DEFAULT 0,
        last_error TEXT,
        next_retry_ts INTEGER DEFAULT 0,
        created_ts INTEGER NOT NULL
      )
    """.trimIndent())
    db.execSQL("CREATE INDEX IF NOT EXISTS idx_sync_queue_retry ON sync_queue(next_retry_ts)")
  }

  override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
    if (oldVersion < 2) {
      db.execSQL("ALTER TABLE locations ADD COLUMN delivered_to_js INTEGER DEFAULT 0")
    }
    if (oldVersion < 3) {
      try { db.execSQL("ALTER TABLE locations ADD COLUMN distance_from_prev REAL DEFAULT 0") } catch (_: Exception) {}
      try { db.execSQL("ALTER TABLE locations ADD COLUMN cumulative_distance REAL DEFAULT 0") } catch (_: Exception) {}
      try { db.execSQL("ALTER TABLE locations ADD COLUMN battery_level REAL DEFAULT -1") } catch (_: Exception) {}
      try { db.execSQL("ALTER TABLE locations ADD COLUMN motion_state TEXT DEFAULT 'unknown'") } catch (_: Exception) {}
      try { db.execSQL("ALTER TABLE locations ADD COLUMN signal_strength TEXT DEFAULT 'medium'") } catch (_: Exception) {}
    }
    if (oldVersion < 4) {
      db.execSQL("""
        CREATE TABLE IF NOT EXISTS native_logs (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          level TEXT NOT NULL,
          message TEXT NOT NULL,
          timestamp INTEGER NOT NULL
        )
      """.trimIndent())
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_native_logs_ts ON native_logs(timestamp)")
    }
    if (oldVersion < 5) {
      db.execSQL("""
        CREATE TABLE IF NOT EXISTS diagnostics_timeline (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          session_id TEXT,
          level TEXT NOT NULL,
          event TEXT NOT NULL,
          data_json TEXT,
          timestamp INTEGER NOT NULL
        )
      """.trimIndent())
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_diag_session_ts ON diagnostics_timeline(session_id, timestamp)")

      db.execSQL("""
        CREATE TABLE IF NOT EXISTS motion_events (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          session_id TEXT,
          activity TEXT NOT NULL,
          confidence REAL DEFAULT 0,
          data_json TEXT,
          timestamp INTEGER NOT NULL
        )
      """.trimIndent())
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_motion_session_ts ON motion_events(session_id, timestamp)")

      db.execSQL("""
        CREATE TABLE IF NOT EXISTS geofence_events (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          session_id TEXT,
          identifier TEXT NOT NULL,
          action TEXT NOT NULL,
          data_json TEXT,
          timestamp INTEGER NOT NULL
        )
      """.trimIndent())
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_geofence_session_ts ON geofence_events(session_id, timestamp)")

      db.execSQL("""
        CREATE TABLE IF NOT EXISTS sync_queue (
          id INTEGER PRIMARY KEY AUTOINCREMENT,
          kind TEXT NOT NULL,
          payload_json TEXT NOT NULL,
          attempts INTEGER DEFAULT 0,
          last_error TEXT,
          next_retry_ts INTEGER DEFAULT 0,
          created_ts INTEGER NOT NULL
        )
      """.trimIndent())
      db.execSQL("CREATE INDEX IF NOT EXISTS idx_sync_queue_retry ON sync_queue(next_retry_ts)")
    }
    if (oldVersion < 6) {
      try { db.execSQL("ALTER TABLE locations ADD COLUMN quality_json TEXT") } catch (_: Exception) {}
    }
  }

  // ─── Locations ─────────────────────────────────────────────────────────────

  fun insert(location: StoredLocation): Boolean {
    val cv = ContentValues().apply {
      put("id", location.id)
      put("latitude", location.latitude)
      put("longitude", location.longitude)
      put("accuracy", location.accuracy)
      put("speed", location.speed)
      put("heading", location.heading)
      put("altitude", location.altitude)
      put("timestamp", location.timestamp)
      put("battery_level", location.batteryLevel)
      put("motion_state", location.motionState)
      put("signal_strength", location.signalStrength)
      put("session_id", location.sessionId)
      put("delivered_to_js", if (location.deliveredToJs) 1 else 0)
      put("distance_from_prev", location.distanceFromPrev)
      put("cumulative_distance", location.cumulativeDistance)
      put("quality_json", location.qualityJson)
    }
    return writableDatabase.insertWithOnConflict("locations", null, cv, SQLiteDatabase.CONFLICT_REPLACE) != -1L
  }

  fun getPendingForJs(limit: Int): List<StoredLocation> {
    val results = mutableListOf<StoredLocation>()
    val cursor = readableDatabase.query(
      "locations", null, "delivered_to_js = 0", null, null, null,
      "timestamp ASC", limit.toString(),
    )
    cursor.use {
      while (it.moveToNext()) {
        results.add(rowToLocation(it))
      }
    }
    return results
  }

  fun getPendingForSession(sessionId: String, limit: Int = 5000): List<StoredLocation> {
    val results = mutableListOf<StoredLocation>()
    val cursor = readableDatabase.query(
      "locations", null, "session_id = ? AND delivered_to_js = 0",
      arrayOf(sessionId), null, null, "timestamp ASC", limit.toString(),
    )
    cursor.use {
      while (it.moveToNext()) {
        results.add(rowToLocation(it))
      }
    }
    return results
  }

  fun markDelivered(ids: List<String>): Int {
    if (ids.isEmpty()) return 0
    val cv = ContentValues().apply { put("delivered_to_js", 1) }
    return ids.sumOf { id ->
      writableDatabase.update("locations", cv, "id = ?", arrayOf(id))
    }
  }

  fun acknowledge(ids: List<String>): Int {
    if (ids.isEmpty()) return 0
    val placeholders = ids.joinToString(",") { "?" }
    return writableDatabase.delete("locations", "id IN ($placeholders)", ids.toTypedArray())
  }

  fun purgeDelivered(): Int {
    return writableDatabase.delete("locations", "delivered_to_js = 1", null)
  }

  fun pendingCount(): Int {
    val cursor = readableDatabase.rawQuery("SELECT COUNT(*) FROM locations WHERE delivered_to_js = 0", null)
    cursor.use { if (it.moveToFirst()) return it.getInt(0) }
    return 0
  }

  fun clearAll() {
    writableDatabase.delete("locations", null, null)
    writableDatabase.delete("sessions", null, null)
  }

  fun getAllLocations(limit: Int = 1000): List<StoredLocation> {
    val results = mutableListOf<StoredLocation>()
    val cursor = readableDatabase.query(
      "locations", null, null, null, null, null,
      "timestamp DESC", limit.toString(),
    )
    cursor.use { while (it.moveToNext()) results.add(rowToLocation(it)) }
    return results
  }

  fun destroyLocation(id: String): Boolean {
    return writableDatabase.delete("locations", "id = ?", arrayOf(id)) > 0
  }

  fun logNative(level: String, message: String) {
    val cv = ContentValues().apply {
      put("level", level)
      put("message", message)
      put("timestamp", System.currentTimeMillis())
    }
    writableDatabase.insert("native_logs", null, cv)
  }

  fun getNativeLog(start: Long?, end: Long?, order: Int, limit: Int): String {
    val clauses = mutableListOf<String>()
    val args = mutableListOf<String>()
    start?.let { clauses.add("timestamp >= ?"); args.add(it.toString()) }
    end?.let { clauses.add("timestamp <= ?"); args.add(it.toString()) }
    val where = if (clauses.isEmpty()) "" else "WHERE ${clauses.joinToString(" AND ")}"
    val orderDir = if (order >= 0) "ASC" else "DESC"
    val cursor = readableDatabase.rawQuery(
      "SELECT level, message, timestamp FROM native_logs $where ORDER BY timestamp $orderDir LIMIT $limit",
      args.toTypedArray(),
    )
    val lines = mutableListOf<String>()
    cursor.use {
      while (it.moveToNext()) {
        lines.add("[${it.getLong(2)}] [${it.getString(0)}] ${it.getString(1)}")
      }
    }
    return lines.joinToString("\n")
  }

  fun destroyNativeLog() {
    writableDatabase.delete("native_logs", null, null)
  }

  fun purgeNativeLogsBefore(cutoffMs: Long) {
    writableDatabase.delete("native_logs", "timestamp < ?", arrayOf(cutoffMs.toString()))
  }

  fun deleteLocationsForSession(sessionId: String) {
    writableDatabase.delete("locations", "session_id = ?", arrayOf(sessionId))
  }

  // ─── Sessions ──────────────────────────────────────────────────────────────

  fun createSession(name: String, activityType: String, extras: String?): String {
    val id = java.util.UUID.randomUUID().toString()
    val cv = ContentValues().apply {
      put("id", id)
      put("name", name)
      put("activity_type", activityType)
      put("start_time", System.currentTimeMillis())
      extras?.let { put("extras", it) }
    }
    writableDatabase.insertWithOnConflict("sessions", null, cv, SQLiteDatabase.CONFLICT_REPLACE)
    return id
  }

  fun endSession(sessionId: String, data: Map<String, Any?>) {
    val cv = ContentValues().apply {
      put("end_time", System.currentTimeMillis())
      put("total_distance", (data["totalDistance"] as? Double) ?: 0.0)
      put("total_duration", (data["totalDuration"] as? Long) ?: 0L)
      put("total_active_duration", (data["totalActiveDuration"] as? Long) ?: 0L)
      put("max_speed", (data["maxSpeed"] as? Double) ?: 0.0)
      put("elevation_gain", (data["elevationGain"] as? Double) ?: 0.0)
      put("average_accuracy", (data["averageAccuracy"] as? Double) ?: 0.0)
      put("point_count", (data["pointCount"] as? Int) ?: 0)
    }
    writableDatabase.update("sessions", cv, "id = ?", arrayOf(sessionId))
  }

  fun discardSession(sessionId: String) {
    deleteLocationsForSession(sessionId)
    writableDatabase.delete("sessions", "id = ?", arrayOf(sessionId))
  }

  fun getUnuploadedSessions(): List<Map<String, Any?>> {
    val results = mutableListOf<Map<String, Any?>>()
    val cursor = readableDatabase.rawQuery(
      "SELECT * FROM sessions WHERE uploaded = 0 ORDER BY start_time ASC", null
    )
    cursor.use {
      while (it.moveToNext()) {
        results.add(sessionRowToMap(it))
      }
    }
    return results
  }

  fun getSessionForUpload(sessionId: String): Map<String, Any?>? {
    val cursor = readableDatabase.rawQuery(
      "SELECT * FROM sessions WHERE id = ?", arrayOf(sessionId)
    )
    cursor.use {
      if (it.moveToFirst()) {
        val sessionData = sessionRowToMap(it).toMutableMap()
        val points = getPendingForSession(sessionId).map { loc ->
          mapOf(
            "id" to loc.id,
            "latitude" to loc.latitude,
            "longitude" to loc.longitude,
            "timestamp" to loc.timestamp,
            "accuracy" to loc.accuracy,
            "speed" to loc.speed,
            "altitude" to loc.altitude,
            "distanceFromPrev" to loc.distanceFromPrev,
            "cumulativeDistance" to loc.cumulativeDistance,
          )
        }
        sessionData["points"] = points
        return sessionData
      }
    }
    return null
  }

  fun markSessionUploaded(sessionId: String) {
    val cv = ContentValues().apply { put("uploaded", 1) }
    writableDatabase.update("sessions", cv, "id = ?", arrayOf(sessionId))
    deleteLocationsForSession(sessionId)
  }

  // ─── Helpers ───────────────────────────────────────────────────────────────

  private fun rowToLocation(c: Cursor): StoredLocation {
    return StoredLocation(
      id = c.getString(c.getColumnIndexOrThrow("id")),
      latitude = c.getDouble(c.getColumnIndexOrThrow("latitude")),
      longitude = c.getDouble(c.getColumnIndexOrThrow("longitude")),
      accuracy = c.getFloat(c.getColumnIndexOrThrow("accuracy")),
      speed = c.getFloat(c.getColumnIndexOrThrow("speed")),
      heading = c.getFloat(c.getColumnIndexOrThrow("heading")),
      altitude = c.getDouble(c.getColumnIndexOrThrow("altitude")),
      timestamp = c.getLong(c.getColumnIndexOrThrow("timestamp")),
      batteryLevel = tryOrNull { c.getDouble(c.getColumnIndexOrThrow("battery_level")) } ?: -1.0,
      motionState = tryOrNull { c.getString(c.getColumnIndexOrThrow("motion_state")) } ?: "unknown",
      signalStrength = tryOrNull { c.getString(c.getColumnIndexOrThrow("signal_strength")) } ?: "medium",
      sessionId = tryOrNull { c.getString(c.getColumnIndexOrThrow("session_id")) } ?: "default",
      deliveredToJs = c.getInt(c.getColumnIndexOrThrow("delivered_to_js")) == 1,
      distanceFromPrev = tryOrNull { c.getDouble(c.getColumnIndexOrThrow("distance_from_prev")) } ?: 0.0,
      cumulativeDistance = tryOrNull { c.getDouble(c.getColumnIndexOrThrow("cumulative_distance")) } ?: 0.0,
      qualityJson = tryOrNull { c.getString(c.getColumnIndexOrThrow("quality_json")) },
    )
  }

  private fun sessionRowToMap(c: Cursor): Map<String, Any?> {
    return mapOf(
      "sessionId" to c.getString(c.getColumnIndexOrThrow("id")),
      "name" to c.getString(c.getColumnIndexOrThrow("name")),
      "activityType" to c.getString(c.getColumnIndexOrThrow("activity_type")),
      "startTime" to c.getLong(c.getColumnIndexOrThrow("start_time")),
      "endTime" to c.getLong(c.getColumnIndexOrThrow("end_time")),
      "totalDistance" to c.getDouble(c.getColumnIndexOrThrow("total_distance")),
      "totalDuration" to c.getLong(c.getColumnIndexOrThrow("total_duration")),
      "totalActiveDuration" to c.getLong(c.getColumnIndexOrThrow("total_active_duration")),
      "maxSpeed" to c.getDouble(c.getColumnIndexOrThrow("max_speed")),
      "elevationGain" to c.getDouble(c.getColumnIndexOrThrow("elevation_gain")),
      "pointCount" to c.getInt(c.getColumnIndexOrThrow("point_count")),
      "pauseCount" to c.getInt(c.getColumnIndexOrThrow("pause_count")),
      "uploaded" to (c.getInt(c.getColumnIndexOrThrow("uploaded")) == 1),
    )
  }

  private fun <T> tryOrNull(block: () -> T): T? = try { block() } catch (_: Exception) { null }
}

File: ios/MicimGeolocation/LocationDatabase.swift
import Foundation
import SQLite3

struct StoredLocation {
  let id: String
  let latitude: Double
  let longitude: Double
  let accuracy: Double
  let speed: Double
  let heading: Double
  let altitude: Double
  let timestamp: Int64
  let batteryLevel: Double
  let signalStrength: String
  let provider: String
  let motionState: String
  let confidence: Double
  let sessionId: String
  let deliveredToJs: Bool
}

final class LocationDatabase {
  static let shared = LocationDatabase()

  private var db: OpaquePointer?
  private let queue = DispatchQueue(label: "com.micim.geolocation.database", qos: .userInitiated)

  private init() {
    openDatabase()
    createTables()
    migrateIfNeeded()
  }

  deinit {
    if db != nil { sqlite3_close(db) }
  }

  private var dbPath: String {
    let dir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first!
    return (dir as NSString).appendingPathComponent("micim_geolocation.db")
  }

  private func openDatabase() {
    if sqlite3_open(dbPath, &db) != SQLITE_OK {
      print("[MicimGeolocation] Failed to open database")
    }
  }

  private func createTables() {
    let sql = """
    CREATE TABLE IF NOT EXISTS locations (
      id TEXT PRIMARY KEY,
      latitude REAL NOT NULL,
      longitude REAL NOT NULL,
      accuracy REAL,
      speed REAL,
      heading REAL,
      altitude REAL,
      timestamp INTEGER NOT NULL,
      battery_level REAL,
      signal_strength TEXT,
      provider TEXT,
      motion_state TEXT,
      confidence REAL,
      session_id TEXT,
      delivered_to_js INTEGER DEFAULT 0
    );
    CREATE INDEX IF NOT EXISTS idx_locations_pending ON locations(delivered_to_js);
    CREATE INDEX IF NOT EXISTS idx_locations_timestamp ON locations(timestamp);
    """
    queue.sync { sqlite3_exec(db, sql, nil, nil, nil) }
  }

  private func migrateIfNeeded() {
    queue.sync {
      var stmt: OpaquePointer?
      let check = "PRAGMA table_info(locations)"
      guard sqlite3_prepare_v2(db, check, -1, &stmt, nil) == SQLITE_OK else { return }
      defer { sqlite3_finalize(stmt) }
      var hasDelivered = false
      while sqlite3_step(stmt) == SQLITE_ROW {
        if let name = sqlite3_column_text(stmt, 1) {
          if String(cString: name) == "delivered_to_js" { hasDelivered = true }
        }
      }
      if !hasDelivered {
        sqlite3_exec(db, "ALTER TABLE locations ADD COLUMN delivered_to_js INTEGER DEFAULT 0", nil, nil, nil)
      }
    }
  }

  /// Write-first: always persist before any JS delivery
  func insert(_ location: StoredLocation) -> Bool {
    queue.sync {
      let sql = """
      INSERT OR REPLACE INTO locations
      (id, latitude, longitude, accuracy, speed, heading, altitude, timestamp,
       battery_level, signal_strength, provider, motion_state, confidence,
       session_id, delivered_to_js)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      """
      var stmt: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return false }
      defer { sqlite3_finalize(stmt) }

      sqlite3_bind_text(stmt, 1, (location.id as NSString).utf8String, -1, nil)
      sqlite3_bind_double(stmt, 2, location.latitude)
      sqlite3_bind_double(stmt, 3, location.longitude)
      sqlite3_bind_double(stmt, 4, location.accuracy)
      sqlite3_bind_double(stmt, 5, location.speed)
      sqlite3_bind_double(stmt, 6, location.heading)
      sqlite3_bind_double(stmt, 7, location.altitude)
      sqlite3_bind_int64(stmt, 8, location.timestamp)
      sqlite3_bind_double(stmt, 9, location.batteryLevel)
      sqlite3_bind_text(stmt, 10, (location.signalStrength as NSString).utf8String, -1, nil)
      sqlite3_bind_text(stmt, 11, (location.provider as NSString).utf8String, -1, nil)
      sqlite3_bind_text(stmt, 12, (location.motionState as NSString).utf8String, -1, nil)
      sqlite3_bind_double(stmt, 13, location.confidence)
      sqlite3_bind_text(stmt, 14, (location.sessionId as NSString).utf8String, -1, nil)
      sqlite3_bind_int(stmt, 15, location.deliveredToJs ? 1 : 0)

      return sqlite3_step(stmt) == SQLITE_DONE
    }
  }

  func markDelivered(ids: [String]) -> Int {
    guard !ids.isEmpty else { return 0 }
    return queue.sync {
      let placeholders = ids.map { _ in "?" }.joined(separator: ",")
      let sql = "UPDATE locations SET delivered_to_js = 1 WHERE id IN (\(placeholders))"
      var stmt: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return 0 }
      defer { sqlite3_finalize(stmt) }
      for (i, id) in ids.enumerated() {
        sqlite3_bind_text(stmt, Int32(i + 1), (id as NSString).utf8String, -1, nil)
      }
      guard sqlite3_step(stmt) == SQLITE_DONE else { return 0 }
      return Int(sqlite3_changes(db))
    }
  }

  /// Locations collected while JS was suspended — replay to watchPosition callbacks
  func getPendingForJs(limit: Int = 200) -> [StoredLocation] {
    queue.sync {
      var results: [StoredLocation] = []
      let sql = "SELECT * FROM locations WHERE delivered_to_js = 0 ORDER BY timestamp ASC LIMIT ?"
      var stmt: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return [] }
      defer { sqlite3_finalize(stmt) }
      sqlite3_bind_int(stmt, 1, Int32(limit))
      while sqlite3_step(stmt) == SQLITE_ROW {
        if let loc = rowToLocation(stmt) { results.append(loc) }
      }
      return results
    }
  }

  func pendingCount() -> Int {
    queue.sync {
      var stmt: OpaquePointer?
      guard sqlite3_prepare_v2(db, "SELECT COUNT(*) FROM locations WHERE delivered_to_js = 0", -1, &stmt, nil) == SQLITE_OK else { return 0 }
      defer { sqlite3_finalize(stmt) }
      if sqlite3_step(stmt) == SQLITE_ROW { return Int(sqlite3_column_int(stmt, 0)) }
      return 0
    }
  }

  func acknowledge(ids: [String]) -> Int {
    guard !ids.isEmpty else { return 0 }
    return queue.sync {
      let placeholders = ids.map { _ in "?" }.joined(separator: ",")
      let sql = "DELETE FROM locations WHERE id IN (\(placeholders))"
      var stmt: OpaquePointer?
      guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return 0 }
      defer { sqlite3_finalize(stmt) }
      for (i, id) in ids.enumerated() {
        sqlite3_bind_text(stmt, Int32(i + 1), (id as NSString).utf8String, -1, nil)
      }
      guard sqlite3_step(stmt) == SQLITE_DONE else { return 0 }
      return Int(sqlite3_changes(db))
    }
  }

  func clearAll() {
    queue.sync { sqlite3_exec(db, "DELETE FROM locations", nil, nil, nil) }
  }

  func purgeDelivered() -> Int {
    queue.sync {
      sqlite3_exec(db, "DELETE FROM locations WHERE delivered_to_js = 1", nil, nil, nil)
      return Int(sqlite3_changes(db))
    }
  }

  private func rowToLocation(_ stmt: OpaquePointer?) -> StoredLocation? {
    guard let stmt = stmt else { return nil }
    return StoredLocation(
      id: String(cString: sqlite3_column_text(stmt, 0)),
      latitude: sqlite3_column_double(stmt, 1),
      longitude: sqlite3_column_double(stmt, 2),
      accuracy: sqlite3_column_double(stmt, 3),
      speed: sqlite3_column_double(stmt, 4),
      heading: sqlite3_column_double(stmt, 5),
      altitude: sqlite3_column_double(stmt, 6),
      timestamp: sqlite3_column_int64(stmt, 7),
      batteryLevel: sqlite3_column_double(stmt, 8),
      signalStrength: String(cString: sqlite3_column_text(stmt, 9)),
      provider: String(cString: sqlite3_column_text(stmt, 10)),
      motionState: String(cString: sqlite3_column_text(stmt, 11)),
      confidence: sqlite3_column_double(stmt, 12),
      sessionId: String(cString: sqlite3_column_text(stmt, 13)),
      deliveredToJs: sqlite3_column_int(stmt, 14) == 1
    )
  }
}

extension StoredLocation {
  func toDictionary() -> [String: Any] {
    [
      "id": id,
      "latitude": latitude,
      "longitude": longitude,
      "timestamp": timestamp,
      "accuracy": accuracy,
      "speed": speed,
      "heading": heading,
      "altitude": altitude,
      "batteryLevel": batteryLevel,
      "signalStrength": signalStrength,
      "provider": provider,
      "motionState": motionState,
      "confidence": confidence,
    ]
  }

  func toPositionDictionary() -> [String: Any] {
    [
      "coords": [
        "latitude": latitude,
        "longitude": longitude,
        "altitude": altitude,
        "accuracy": accuracy,
        "heading": heading,
        "speed": speed,
      ],
      "timestamp": timestamp,
    ]
  }
}
